iothub/device/src/InternalClient.cs

```diff
@@ -1186,16 +1184,16 @@ internal void OnConnectionStateChanged(ConnectionState state, ConnectionStateCha
                 if (Logging.IsEnabled)
                     Logging.Enter(this, state, reason, nameof(OnConnectionStateChanged));
 
-                if (_lastConnectionState != state
-                    || _lastConnectionStateChangeReason != reason)
+                if (_connectionInfo.State != state
+                    || _connectionInfo.ChangeReason != reason)
                 {
                     _connectionStateChangeHandler?.Invoke(state, reason);
                 }
             }
             finally
             {
-                _lastConnectionState = state;
-                _lastConnectionStateChangeReason = reason;
+                _connectionInfo = new ConnectionInfo(state, reason, DateTime.UtcNow);
```

Comment:
Set properties in `ConnectionInfo` at the same time so no timing issues if someone checked between assignment of multiple properties.